File: Programmers/Lv. 3/퍼즐 조각 채우기.py
from collections import deque

def solution(game_board, table):
    l = len(table)
    
    moves = [(1,0), (-1,0), (0,1), (0,-1)]
    
    def finding(arr, num):
        visited = [[False] * l for _ in range(l)]
        shapes = []
        for y in range(l):
            for x in range(l):          
                if arr[y][x] == num and not visited[y][x]:
                    q = deque()
                    q.append([y, x])
                    visited[y][x] = True
                    shape = []
                    while q:
                        cy, cx = q.popleft()
                        shape.append([cy, cx])
                        for move in moves:
                            dy = cy + move[0]
                            dx = cx + move[1]
                            if 0 <= dy < l and 0 <= dx < l and not visited[dy][dx]:
                                visited[dy][dx] = True
                                if arr[dy][dx] == num:
                                    q.append([dy, dx])
                    shapes.append(moveToZero(sorted(shape)))
        return shapes
    
    def moveToZero(arr):
        min_y, min_x = 50, 50
        tmp = []
        for y, x in arr:
            min_y = min(min_y, y)
            min_x = min(min_x, x)
        for y, x in arr:
            tmp.append([y - min_y, x - min_x])
        return tmp
    
    def rotation(arr):
        result = [arr]
        for _ in range(3):
            tmp = []
            for y, x in arr:
                tmp.append([x, -y])
            arr = moveToZero(sorted(tmp))
            result.append(arr)
        return result
            
    # table에서 블록 찾고 (0,0) 으로 이동
    blocks = finding(table, 1)
    # game_board에서 빈 곳 찾고 (0,0) 으로 이동
    spaces = finding(game_board, 0)
        
    answer = 0
    # 맞는 블록 찾기    
    q = deque()
    used = [i for i in range(len(blocks))]
    filled = [i for i in range(len(spaces))]
    q.append([used, filled])
    while q:
        _used, _filled = q.popleft()
        findFlag = False
        for i in _used:
            for j in _filled:
                if spaces[j] in rotation(blocks[i]):
                    _used.remove(i)
                    _filled.remove(j)
                    q.append([_used, _filled])
                    answer += len(blocks[i])
                    findFlag = True
                    break          
            if findFlag:
                break
                                       
    return answer

File: Programmers/Lv. 3/퍼즐 조각 채우기.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int y_move[4] = {1, -1, 0, 0};
int x_move[4] = {0, 0, 1, -1};

vector<pair<int, int>> MoveToZero(vector<pair<int, int>> v)
{
    int min_Y = 51;
    int min_X = 51;
    
    for (pair<int, int> p : v)
    {
        if (p.first < min_Y)
        {
            min_Y = p.first;
        }
        if (p.second < min_X)
        {
            min_X = p.second;
        }
    }
        
    vector<pair<int, int>> new_v;
    sort(v.begin(), v.end());
    for (pair<int, int> p : v)
    {
        new_v.push_back({p.first - min_Y, p.second - min_X});
    }
    
    return new_v;
}

vector<vector<pair<int, int>>> FindShapes(vector<vector<int>> board, int num)
{
    vector<vector<pair<int, int>>> results;
    
    int size_Y = board.size();
    int size_X = board[0].size();
    
    bool visited[51][51] = {false, };
    
    
    for (int y = 0; y < size_Y; y++)
    {
        for (int x = 0; x < size_X; x++)
        {
            queue<pair<int, int>> q;       
            vector<pair<int, int>> v;     
            
            if (!visited[y][x] && board[y][x] == num)
            {
                visited[y][x] = true;
                
                v.push_back({y, x});
                q.push({y, x});
                
                while (!q.empty())
                {
                    auto tmp = q.front();
                    q.pop();
                    
                    int cy = tmp.first;
                    int cx = tmp.second;
                    
                    for (int i = 0; i < 4; i++)
                    {
                        int ny = cy + y_move[i];
                        int nx = cx + x_move[i];
                        
                        if (ny >= 0 && ny < size_Y && nx >= 0 && nx < size_X && !visited[ny][nx] && board[ny][nx] == num)
                        {
                            visited[ny][nx] = true;
                            
                            v.push_back({ny, nx});
                            
                            q.push({ny, nx});
                        }
                    }
                }                
                results.push_back(MoveToZero(v));
            }
        }
    }
    return results;
}

vector<vector<pair<int, int>>> RotateShapes(vector<pair<int, int>> board)
{
    vector<vector<pair<int, int>>> results;
    results.push_back(board);
    
    for (int i = 0; i < 3; i++)
    {
        vector<pair<int, int>> tmp_v;
        for (pair<int, int> p : board)
        {
            tmp_v.push_back({p.second, -p.first});
        }
        board = MoveToZero(tmp_v);
        results.push_back(board);
    }
    
    return results;
}

int solution(vector<vector<int>> game_board, vector<vector<int>> table) {
    int answer = 0;
    
    vector<vector<pair<int, int>>> emptys = FindShapes(game_board, 0);
    vector<vector<pair<int, int>>> blocks = FindShapes(table, 1);
    
    vector<bool> used;
    for (int i = 0; i < blocks.size(); i++)
    {
        used.push_back(false);
    }
    
    for (int i = 0; i < emptys.size(); i++)
    {
        for (int j = 0; j < blocks.size(); j++)
        {            
            auto rotated = RotateShapes(blocks[j]);
            if (!used[j] && count(rotated.begin(), rotated.end(), emptys[i]) != 0)
            {               
                used[j] = true;
                answer += emptys[i].size();
                break;
            }
        }
    }
        
    return answer;
}